src/sum.ts

```diff
@@ -14,19 +14,33 @@ import { isAsyncIterable, isIterable } from "./_internal/utils";
  * await sum(toAsync(['a', 'b', 'c'])); // 'abc'
  * ```
  */
-function sum<
-  A extends
-    | Iterable<number>
-    | AsyncIterable<number>
-    | Iterable<string>
-    | AsyncIterable<string>,
->(iterable: A): ReturnValueType<A> {
+function sum<A extends readonly []>(iterable: A): 0;
+function sum<A extends Iterable<number> | AsyncIterable<number>>(
+  iterable: A,
+): ReturnValueType<A>;
+
+function sum<A extends Iterable<number> | AsyncIterable<number> | readonly []>(
+  iterable: A,
+): ReturnValueType<A> | 0 {
   if (isIterable(iterable)) {
+    const iterator = iterable[Symbol.iterator]();
+    const { done } = iterator.next();
+    if (done) {
+      return 0;
+    }
+
     return reduce(add, iterable) as ReturnValueType<A>;
   }
 
   if (isAsyncIterable(iterable)) {
-    return reduce(add, iterable) as ReturnValueType<A>;
+    const iterator = iterable[Symbol.asyncIterator]();
+    return iterator.next().then(({ done }) => {
+      if (done) {
+        return 0;
+      }
+
+      return reduce(add, iterable) as ReturnValueType<A>;
+    });
```

Comment:
How to fix this type error...?

<img width="966" alt="image" src="https://user-images.githubusercontent.com/10627668/184499483-07f9835f-e965-4a4e-8faf-7d16ee026d2a.png">

Reply:
Thank you for your contribution👍

I left a comment
When we check the return type of `reduce`, we can see that it does not return as we intended.

<img width="935" alt="image" src="https://user-images.githubusercontent.com/10924072/184544367-532f8518-ab1f-4a3a-a557-abd2d9369907.png">

So, if you pass the type as a generic to `reduce`, you can avoid an error.

```ts
   const iterator = iterable[Symbol.asyncIterator]();
     return iterator.next().then(({ done }) => {
       if (done) {
         return 0;
       }

       return reduce<number, number>(add, iterable);
     }) as ReturnValueType<A>;
```

However, if you do this, you can see that type inference does not work well in `sum.test.ts`
I think we need to tell TypeScript a little more about types.

It seems that the singature should be able to receive only the number as shown below.
```ts
function sum<A extends Iterable<number> | AsyncIterable<number>>(
   iterable: A,
): A extends Iterable<number>
   ? number
   : A extends AsyncIterable<number>
   ? Promise<number>
   : never;
```

we decided to take only number, I don't think we need to use `ReturnValueType`.
And the type we will actually implement is as follows.
```ts
function sum<A extends Iterable<number> | AsyncIterable<number>>(
   iterable: A,
): number | Promise<number>
```


Additionally, since we've checked the first value once, we don't need to check the first value again.
```ts
 const iterator = iterable[Symbol.iterator]();
 const { done, value } = iterator.next();
 if (done) {
    return 0;
 }

 return reduce<number, number>((a, b) => add(a, b), value, {
   [Symbol.iterator]() {
      return iterator;
   },
 });
```

Reply:
How about this?
The reduce function is already processing empty values.
```ts
function sum<A extends readonly []>(iterable: A): 0;
function sum<A extends number[]>(arr: A): number;
function sum<A extends Iterable<number>>(iterable: A): number;
function sum<A extends AsyncIterable<number>>(iterable: A): Promise<number>;

function sum<A extends Iterable<number> | AsyncIterable<number>>(
  iterable: A,
): number | Promise<number> {
  if (Array.isArray(iterable)) {
    return iterable.reduce(add, 0);
  } else if (isIterable(iterable)) {
    return reduce<number, number>(add, 0, iterable);
  } else if (isAsyncIterable(iterable)) {
    return reduce<number, number>(add, Promise.resolve(0), iterable);
  }

  throw new TypeError("'iterable' must be type of Iterable or AsyncIterable");
}
```

Reply:
Thank you for feedback 🍺

Reply:
https://github.com/marpple/FxTS/pull/169/commits/02d23eed89df97ed88a7c66041b7a8205f652bcc